src/rgw/rgw_aws_sns.cc

```diff
@@ -0,0 +1,86 @@
+#include <boost/lockfree/queue.hpp>
+#include <string>
+#include <utility>
+#include <aws/sns/SNSClient.h>
+#include <aws/sns/model/PublishRequest.h>
+
+namespace sns {
+
+    static const int STATUS_OK = 1;
+    static const int STATUS_FULL = 2;
+    static const int STATUS_MANAGER_STOPPED = 3;
+
+    struct message_t {
+        std::string topic;
+        std::string message;
+
+        message_t(std::string topic, std::string message) : topic(std::move(topic)), message(std::move(message)) {};
+    };
+
+    typedef boost::lockfree::queue<message_t *, boost::lockfree::fixed_sized<true>> MessageQueue;
+
+    class Manager {
+    private:
+        MessageQueue messageQueue;
+        std::thread runner;
+        bool stopped;
+
+        void publish_internal(message_t *msg) {
+          std::cout << "Publishing.." << std::endl;
+          Aws::SNS::Model::PublishRequest req;
+          req.SetMessage(Aws::String(msg->message));
+          req.SetTopicArn(Aws::String(msg->topic));
+          Aws::SNS::SNSClient client;
+
+          auto result = client.Publish(req);
+          if (result.IsSuccess()) {
+            std::cout << "Published..." << std::endl;
+          } else {
+            std::cout << result.GetError().GetMessage() << std::endl;
+          }
+        }
+
+//    Works on messages in the queque
+        void run() {
+          while (!stopped) {
+            const auto count = messageQueue.consume_all(
```

Comment:
in case that the queue is empty the thread will be spinning at 100% CPU.
please add a sleep command if idle. see an example [here](https://github.com/ceph/ceph/blob/master/src/rgw/rgw_amqp.cc#L792)